cardano-cli/src/Cardano/CLI/EraBased/HashChecking.hs

```diff
@@ -0,0 +1,123 @@
+          case shelleyPoolCert of
+            L.RegPool poolParams -> forM_ (L.ppMetadata poolParams) checkPoolMetadataHash
+            L.RetirePool _ _ -> return ()
+        L.ShelleyTxCertGenesisDeleg _ -> return ()
+        L.ShelleyTxCertMir _ -> return ()
+    ConwayCertificate ceo conwayCert ->
+      Shelley.conwayEraOnwardsConstraints ceo $
+        case conwayCert of
+          L.ConwayTxCertDeleg _ -> return ()
+          L.ConwayTxCertPool conwayPoolCert ->
+            case conwayPoolCert of
+              L.RegPool poolParams -> forM_ (L.ppMetadata poolParams) checkPoolMetadataHash
+              L.RetirePool _ _ -> return ()
+          L.ConwayTxCertGov govCert ->
+            case govCert of
+              L.ConwayRegDRep _ _ mAnchor -> forM_ mAnchor checkAnchorMetadataHash
+              L.ConwayUnRegDRep _ _ -> return ()
+              L.ConwayUpdateDRep _ mAnchor -> forM_ mAnchor checkAnchorMetadataHash
+              L.ConwayAuthCommitteeHotKey _ _ -> return ()
+              L.ConwayResignCommitteeColdKey _ mAnchor -> forM_ mAnchor checkAnchorMetadataHash
+ where
+  checkPoolMetadataHash :: L.PoolMetadata -> ExceptT TxCmdError IO ()
+  checkPoolMetadataHash (L.PoolMetadata{L.pmUrl = url, L.pmHash = hashBytes}) = do
+    let mHash = L.hashFromBytes hashBytes
+    hash <- maybe (left $ TxCmdPoolMetadataHashError url) return mHash
+    let safeHash = L.unsafeMakeSafeHash hash
+    checkAnchorMetadataHash
+      ( L.Anchor
+          { L.anchorUrl = url
+          , L.anchorDataHash = safeHash
+          }
+      )
+
+-- | Find references to anchor data in voting procedures and check the hashes are valid
+-- and they match the linked data.
+checkVotingProcedureHashes
+  :: Shelley.ShelleyBasedEra era -> Shelley.VotingProcedures era -> ExceptT TxCmdError IO ()
+checkVotingProcedureHashes eon (Shelley.VotingProcedures (L.VotingProcedures voterMap)) =
+  Shelley.shelleyBasedEraConstraints eon $
+    forM_
+      voterMap
+      ( \vpMap ->
+          forM_
+            vpMap
+            ( \(L.VotingProcedure _ mAnchor) ->
+                forM_ mAnchor checkAnchorMetadataHash
+            )
+      )
+
+-- | Find references to anchor data in proposals and check the hashes are valid
+-- and they match the linked data.
+checkProposalHashes
+  :: forall era. Shelley.ShelleyBasedEra era -> Shelley.Proposal era -> ExceptT TxCmdError IO ()
+checkProposalHashes
+  eon
+  ( Shelley.Proposal
+      ( L.ProposalProcedure
+          { L.pProcGovAction = govAction
+          , L.pProcAnchor = anchor
+          }
+        )
+    ) =
+    Shelley.shelleyBasedEraConstraints eon $ do
+      checkAnchorMetadataHash anchor
+      checkGovActionHashes govAction
+   where
+    checkGovActionHashes
+      :: L.GovAction (Shelley.ShelleyLedgerEra era) -> ExceptT TxCmdError IO ()
+    checkGovActionHashes govAction' =
+      Shelley.shelleyBasedEraConstraints eon $
+        case govAction' of
+          L.ParameterChange{} -> return ()
```

Comment:
A comment stating that only the `NewConstitution` governance action contains a checkable hash would be useful.

Reply:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/951/commits/78f05e0d79706772760b447d976ff27b59f47763